File: Labs/Lab 1/TwoSum.java
import java.util.HashMap;

public class TwoSum {
  public static int[] twoSum(int[] nums, int target) {
    HashMap<Integer, Integer> map = new HashMap<>();

  for (int i = 0; i < nums.length(); i++) {
      int complement = target - nums[i];
    // Check if complement exists 
    if (map.containsKey(a)) {
        return new int[]{map.get(a),i};
      }
    // if not add cur element and its index to map
    map.put(nums[i], i);
  }
    //No solution found
  throw new IllegalArgumentException("No two sum solution");
  }

public static void main(String[] args) {
  int[] nums = {3,5,10,14};
  int target = 8;
  int[] result = twoSum(nums,target);
  System.out.println("Indices: " + result[0] + ", " + result[1]);
}

File: Labs/Lab 1/TwoSum.py
def twoSum(nums, target):
    # Using a dictionary to store the difference between target and each element
    num_dict = {}

    for i, num in enumerate(nums):
        complement = target - num

        # Check if the complement exists in the dictionary
        if complement in num_dict:
            return [num_dict[complement], i]

        # If not, add the current element and its index to the dictionary
        num_dict[num] = i

    # No solution found, return an empty list or raise an exception based on requirements
    raise ValueError("No two sum solution")

# Test 
nums = [2, 7, 11, 15]
target = 9
result = twoSum(nums, target)
print(f"Indices: {result[0]}, {result[1]}")
